src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/documents/DefendantResponseReceiptServiceIntegrationTest.java

```diff
@@ -0,0 +1,98 @@
+import org.junit.runner.RunWith;
+import org.mockito.stubbing.Answer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.test.context.junit4.SpringRunner;
+import uk.gov.hmcts.cmc.claimstore.BaseMockSpringTest;
+import uk.gov.hmcts.cmc.claimstore.config.properties.pdf.DocumentTemplates;
+import uk.gov.hmcts.cmc.claimstore.documents.content.DefendantResponseContentProvider;
+import uk.gov.hmcts.cmc.claimstore.helper.DocumentComparisonHelper;
+import uk.gov.hmcts.cmc.claimstore.helper.HTMLTemplateProcessor;
+import uk.gov.hmcts.cmc.domain.models.Claim;
+import uk.gov.hmcts.cmc.domain.models.response.DefenceType;
+import uk.gov.hmcts.cmc.domain.models.sampledata.SampleClaim;
+import uk.gov.hmcts.cmc.domain.models.sampledata.SampleResponse;
+import uk.gov.hmcts.cmc.email.EmailService;
+import uk.gov.hmcts.reform.pdf.service.client.PDFServiceClient;
+
+import java.time.LocalDateTime;
+import java.util.Map;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+import static uk.gov.hmcts.cmc.claimstore.helper.FileUtils.readFile;
+import static uk.gov.hmcts.cmc.claimstore.helper.FileUtils.writeFile;
+import static wiremock.org.custommonkey.xmlunit.XMLAssert.assertXMLEqual;
+
+@RunWith(SpringRunner.class)
+public class DefendantResponseReceiptServiceIntegrationTest extends BaseMockSpringTest {
+
+    @Autowired
+    private DefendantResponseReceiptService defendantResponseReceiptService;
+
+    @Autowired
+    protected DefendantResponseContentProvider contentProvider;
+
+    @Autowired
+    protected DocumentTemplates documentTemplates;
+
+    @Autowired
+    protected DocumentComparisonHelper documentComparisonHelper;
+
+    @Autowired
+    protected HTMLTemplateProcessor htmlTemplateProcessor;
+
+    @MockBean
+    protected EmailService emailService;
+
+    @Before
+    public void beforeEachTest() {
+
+        provideLocalPdfService();
+    }
+
+    @Test
+    public void shouldGenerateDefenceResponseForClaimWithFullDefenceAlreadyPaid() throws Exception {
+
+        Claim claim = SampleClaim.builder()
+            .withResponse(
+                SampleResponse.FullDefence
+                    .builder()
+                    .withDefenceType(DefenceType.ALREADY_PAID)
+                    .withMediation(null)
+                    .build()
+            )
+            .withRespondedAt(LocalDateTime.now())
+            .build();
+
+        byte[] actualHtmlBytes = defendantResponseReceiptService.createHtml(claim);
+
+        String actualHtml = documentComparisonHelper.replaceTimestamp("timestamp",
+            new String(actualHtmlBytes));
+
+        writeFile("build/tmp/actual.html", actualHtml); // Useful file for debugging test
+
+        String expectedHtml = readFile("src/integrationTest/resources/documents/expectedDefendantResponseReceipt.html");
+
+        assertXMLEqual(actualHtml, expectedHtml);
```

Comment:
I'm not sure this is the right approach; we'll have to mirror any changes to the template in the expected document. The test shouldn't fail just because we changed a font size or switched from `<ol>` to `<ul>`.

Instead, I suggest searching the document for appropriate text according to the input. In this case, you created a full defence, already paid; so to verify it we'd look for "I have paid what I believe I owe". Then you could add a couple more scenarios covering the high-level variations, and we won't need the xmlAssert dependency.

Let me know what you think!

Reply:
I could use a simple text equals comparison instead of xmlAssert. However, the advantages of xmlAssert are that it will validate that the document is well formed, and it will identify the location of any differences using XPath. 

You are right that doing a HTML document comparison will mean that some trivial changes will cause test failures. But it also means that it will detect formatting regressions that would not be found by doing specific text comparisons. It is also is an easy way to get full coverage of the document content.